src/raft.rs

```diff
@@ -1966,6 +1973,47 @@ impl<T: Storage> Raft<T> {
         Ok(())
     }
 
+    /// Commits the logs using commit info in vote message.
+    fn maybe_commit_by_vote(&mut self, m: &Message) {
+        if m.commit == 0 || m.commit_term == 0 {
+            return;
+        }
+        let last_commit = self.raft_log.committed;
+        if m.commit <= last_commit || self.state == StateRole::Leader {
+            return;
+        }
+        if !self.raft_log.maybe_commit(m.commit, m.commit_term) {
+            return;
+        }
+
+        let log = &mut self.r.raft_log;
+        info!(self.r.logger, "[commit: {}, lastindex: {}, lastterm: {}] fast-forwarded commit to vote request [index: {}, term: {}]",
+                log.committed, log.last_index(), log.last_term(), m.commit, m.commit_term);
+
+        if self.state != StateRole::Candidate && self.state != StateRole::PreCandidate {
+            return;
+        }
+
+        let ents = self
+            .raft_log
+            .slice(last_commit + 1, self.raft_log.committed + 1, None)
+            .unwrap_or_else(|e| {
+                fatal!(
+                    self.logger,
+                    "unexpected error getting unapplied entries [{}, {}): {:?}",
+                    last_commit + 1,
+                    self.raft_log.committed + 1,
+                    e
+                );
+            });
+        if self.num_pending_conf(&ents) != 0 {
+            // The candidate doesn't have to step down in theory, here just for best
+            // safety as we assume quorum won't change during election.
+            let term = self.term;
+            self.become_follower(term, INVALID_ID);
+        }
+    }
```

Comment:
How about adding a test for this logic?

Reply:
Following assert is exactly for that purpose:

https://github.com/tikv/raft-rs/pull/411/files#diff-fbf583b45a1333793ec034c2c5f932b3d13543758165cf96878a110f56204b84R4732-R4737